src/llamafactory/data/loader.py

```diff
@@ -143,21 +143,21 @@ def _get_merged_dataset(
     data_args: "DataArguments",
     training_args: "Seq2SeqTrainingArguments",
     stage: Literal["pt", "sft", "rm", "ppo", "kto"],
+    merge: bool = True
 ) -> Optional[Union["Dataset", "IterableDataset"]]:
-    r"""
-    Gets the merged datasets in the standard format.
-    """
     if dataset_names is None:
         return None
 
-    datasets = []
+    datasets = {}
     for dataset_attr in get_dataset_list(dataset_names, data_args.dataset_dir):
         if (stage == "rm" and dataset_attr.ranking is False) or (stage != "rm" and dataset_attr.ranking is True):
             raise ValueError("The dataset is not applicable in the current training stage.")
 
-        datasets.append(_load_single_dataset(dataset_attr, model_args, data_args, training_args))
-
-    return merge_dataset(datasets, data_args, seed=training_args.seed)
+        datasets[f'{dataset_attr.dataset_name}_{dataset_attr.subset}'] = _load_single_dataset(dataset_attr, model_args, data_args, training_args)
+    if merge:
+        return merge_dataset([data for _, data in datasets.items()], data_args, seed=training_args.seed)
+    else:
+        return datasets
```

Comment:
does `dict` match the return value type `Optional[Union["Dataset", "IterableDataset"]]`?

Reply:
Oh, It's not match, when merged is False, the  return value type is Dict[str, Dataset].
But `_get_merged_dataset` seems only used in `_get_preprocessed_dataset`， in `_get_preprocessed_dataset` method,Different types  result from `_get_merged_dataset` are treated differently.

I think this code has the least amount of change.
If we want to keep the semantic consistency of function signatures, perhaps we can change the signature of the inner method `_get_merged_dataset` to` _get_merged_optional_dataset`